docs: rfc-0240 XTR-BTC atomic swap (#3485)
tari-project/tari: RFC/src/RFC-0240_AtomicSwap.md

> # RFC-0240/Atomic Swap

Reviewer: ```suggestion
# RFC-0240/Atomic Swaps
```

SERVICES; LOSS OF USE, DATA OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals

> The aim of this Request for Comment (RFC) is to describe how an Atomic swap will be created.

Reviewer: ```suggestion
The aim of this Request for Comment (RFC) is to describe how BTC/XTR Atomic swaps will be created.
```

Author: Specifically avoided calling this a BTC/XTR Atomic swap document so that we can expand this to include all other crypto atomics swaps here.

Reviewer: That’s fine but at the moment that’s all it is 🤷

## Related Requests for Comment

* [RFC-0201: TariScript](RFC-0201_TariScript.md)
* [RFC-0202: TariScript Opcodes](RFC-0202_TariScriptOpcodes.md)

$$
\newcommand{\preimage}{\\phi} % pre image
\newcommand{\hash}[1]{\mathrm{H}\bigl({#1}\bigr)}
$$

## Description

> Atomic swaps are atomic transactions that allow users to trustlessly exchange different crypto assets and or coins without the use of a central exchange. This makes it much more private and secure to do and swap because no third party is required for this to be secure. Atomic swaps work on the principle of using [Hashed Time Lock Contract](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts)(HTLC). In short, it requires some hash pre-image to unlock the contract or the time-lock can be used to reclaim the funds. 

Reviewer: ```suggestion
Atomic swaps are atomic transactions that allow users to trustlessly exchange different crypto assets and or coins without the use of a central exchange. This makes it much more private because no third party is required. Atomic swaps work by using [Hashed Time Lock Contracts](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts)(HTLC). In short, it requires some hash pre-image to unlock the contract, or the time-lock can be used to reclaim the funds. 
```

> In an Atomic swap, both users lock up the funds to be exchanged on their respective chains in an HTLC-type contract. However, the pre-image, or spending secret, of the two contracts is the same, but only one party knows the correct pre-image. When the first HTLC contract is spent, then this publicly reveals the pre-image for the other party to use in spending the second HTLC. If the first HTLC is never spent, the second transaction's time-lock will allow the user to respend the funds back to themselves after a time.

Reviewer: Is it "publicly" revealed? It is revealed to the counterparty I presume. Maybe some clarification about "after a time" as well: how much time? is it the length of the time lock? can it only be spent after that? 

```suggestion
In an Atomic swap, both users lock up the funds to be exchanged on their respective chains in an HTLC-type contract. The pre-image, or spending secret, of the two contracts is the same, but is only known to the initiator. When the first HTLC is spent, then this reveals the pre-image for the other party to use in spending the second HTLC. If the first HTLC is never spent, the second transaction's time-lock will allow the user to spend the funds back to themselves after a time.
```

Author: Yes, its publicly revealed on the blockchain. 
But yeah the time is the time lock, and there is no limit to what it should be or what a minimum is. It is completely user-defined.

### BTC - XTR AtomicSwap

#### Overview

> BTC uses a scripting language to enable smart contracts on transactions which enables [atomic swaps](https://tlu.tarilabs.com/protocols/atomic-swaps/AtomicSwaps.html) on the BTC chain. Traditionally [Mimblewimble] coins do not implement scripts, this makes Atomic swaps harder to implement but not impossible. Grin has implemented atomic swaps as a version of a 2-of-2 multi-signature transaction [mimblewimble atomic swaps](https://tlu.tarilabs.com/protocols/grin-protocol-overview/MainReport.html#atomic-swaps). Fortunately, Tari does have scripting with [TariScript] which works a lot like BTC scripts, this makes adding in scripts a lot easier. Because of the same type of scripting, the scripts to both HTLCs will look very similar and we only need to ensure that we use the same Hashing function in both. 

Reviewer: ```suggestion
BTC uses a scripting language for smart contracts on transactions which enables [atomic swaps](https://tlu.tarilabs.com/protocols/atomic-swaps/AtomicSwaps.html) on the BTC chain. Traditionally [Mimblewimble] coins do not implement scripts, which makes Atomic swaps harder to implement but not impossible. Grin has implemented atomic swaps using a version of a 2-of-2 multi-signature transaction [mimblewimble atomic swaps](https://tlu.tarilabs.com/protocols/grin-protocol-overview/MainReport.html#atomic-swaps). Fortunately, Tari does have scripting with [TariScript] which works a lot like BTC scripts, which make the implementation simpler. Because of the scripting similarities, the scripts to both HTLCs will look very similar and we only need to ensure that we use the same hash function in both. 
```

> To do an Atomic swap from BTC to XTR we need 4 wallets, 2 BTC wallets, and 2 XTR wallets.

Reviewer: I'd suggest clarifying where these numbers come from. Is it 1 of each type of wallet for each party?

Author: yip, will clarify

> As an example, Alice wants to trade some of her XTR for Bob's BTC. Alice and Bob need to agree on an amount of XTR and BTC to swap. If this is agreed on the following steps will happen:

Reviewer: ```suggestion
As an example, Alice wants to trade some of her XTR for Bob's BTC. Alice and Bob need to agree on an amount of XTR and BTC to swap. Once an agreement is reached, the swap is executed in the following steps:
```

> * Alice chooses a random string as the pre-image (\\( \preimage \\)), she hashes this with SHA256. She then sends the hash to Bob. Along with het BTC address.

Reviewer: How is this sent, out-of-band? or is that just an implementation detail?

```suggestion
* Alice chooses a random string as the pre-image (\\( \preimage \\)), she hashes this with SHA256. She then sends the hash to Bob. Along with her BTC address.
```

Author: Yes, implementation detail.

* Bob sends his a public version of the [script key], \\( K_{Sb} \\) for use in the XTR transaction.

> * Alice creates a XTR transaction with an HTLC contract requiring \\( \preimage \\) as the input, to pay out to Bob's address or it pays back to her after a certain time. 

Reviewer: ```suggestion
* Alice creates a XTR transaction with an HTLC contract requiring \\( \preimage \\) as the script input, to pay out to Bob's address or it pays back to her after a certain time. 
```

* Bob waits for this transaction to be mined. When it is mined, he checks the spending script and verifies that the script checks (\\ \hash{\preimage} \\), he has the [script key], \\( K_{Sb} \\) to claim it and that it has a sufficiently long time-lock to give him time to claim the transaction.

> * Upon verification, Bob creates a BTC transaction with the same \\( \preimage \\). But this HTLC spends to Alice's BTC address she gave him. 

Reviewer: This could use some more detail, does the tx have an identical HTLC?

```suggestion
* Upon verification, Bob creates a BTC transaction with a HTLC with the same \\( \preimage \\). But this HTLC spends to Alice's provided BTC address. 
```

> * Alice checks the Bitcoin blockchain and upon seeing that the transaction is mined, she claims the transaction. But by claiming it, she makes public what (\\( \preimage \\)) is.

Reviewer: Is it public? This could use a bit more detail about how the preimage is revealed

> * Bob sees that his BTC is spent, and looks at the witness to get \\( \preimage \\). Bob can then use \\( \preimage \\) to claim the XTR transaction.

Reviewer: This is the first mention of "witness". Is this the witness proof? This could use more detail.

> * Bob sees that his BTC is spent, and looks at the witness to get \\( \preimage \\). Bob can then use \\( \preimage \\) to claim the XTR transaction.

Reviewer 2: ```suggestion
As an example, Alice wants to trade some of her XTR for Bob's BTC. Alice and Bob need to agree on an amount of XTR and 
BTC to swap. If this is agreed on the following steps will happen:

* Alice chooses a random string, \\( \preimage \\), as the pre-image and hashes it with SHA256. She then sends the hash 
of the pre-image, \\( \hash{\preimage} \\), to Bob along with her BTC address.

* Bob sends her a public version of his [script key], \\( K_{Sb} \\), for use in the XTR transaction, which we can refer 
to as Bob's script address.

* Alice creates a one-sided XTR transaction with an HTLC contract requiring \\( \preimage \\) as the input, which will 
either pay out to Bob's script address or to her script address, \\( K_{Sa} \\), after a certain "time" has elapsed 
(block height has been reached). 

* Bob waits for this transaction to be mined. When it is mined, he verifies that the UTXO spending script expects a 
comparison of \\( \hash{\preimage} \\) as the first instruction, and that his public [script key], \\( K_{Sb} \\), will 
be the final value remaining after executing the script. He has the private [script key], \\( k_{Sb} \\), to enable him 
to produce a signature to claim the funds if he can get hold of the expected pre-image input value, \\( \preimage \\). 
He also verifies that the UTXO has a sufficiently long time-lock to give him time to claim the transaction.

* Upon verification, Bob creates a Segwit HTLC BTC transaction with the same \\( \hash{\preimage} \\), which will spend 
to Alice's BTC address she gave him. It is important that the time lock for this HTLC has to expire prior to the time 
lock of the XTR HTLC that Alice created.

* Alice checks the Bitcoin blockchain and upon seeing that the transaction is mined, she claims the transaction, but, 
in order for her to do so, she has to make public what \\( \preimage \\) is as she has to use it as the witness of the 
claiming transaction.

* Bob sees that his BTC is spent, and looks at the witness to get \\( \preimage \\). Bob can then use \\( \preimage \\) 
to claim the XTR transaction.
```

#### BTC HTLC script 

Here is the required BTC script that Bob publishes:

```
	OP_IF
	   OP_SHA256 <HASH256{pre_image}> OP_EQUALVERIFY
		<Alice BTC address> OP_CHECKSIG
	OP_ELSE
      <relative locktime>
      OP_CHECKSEQUENCEVERIFY
      OP_DROP
      <Bob BTC address> OP_CHECKSIG
   OP_ENDIF
```
_relative locktime_ is a time sequence that Alice chooses to lock up the funds in order to give Bob time to claim this. 

> #### XTR - HTLC script 

Reviewer: these 2 headings should have the same formatting, either both with the dash or both without

Reviewer 3: Agree with this, lets make them consistent.

Here is the required XTR script that Alice publishes:

```
   HashSha256 PushHash(HASH256{pre_image}) Equal
   IFTHEN
      PushPubkey(K_{Sb})
	ELSE
      CheckHeightVerify(height)
      PushPubkey(K_{Sa})
   ENDIF
```
(\\( K_{Sb} \\)) is the public key of the [script key] pair that Bob chooses to be able to claim this transaction if Alice backs out. 

> _height_ is a absolute block height that Bob chooses to lock up the funds in order to give Alice time to claim the funds. 

Reviewer: ```suggestion
_height_ is an absolute block height that Bob chooses to lock up the funds in order to give Alice time to claim the funds. 
```

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to describe how an Atomic swap will be created.

## Related Requests for Comment

* [RFC-0201: TariScript](RFC-0201_TariScript.md)
* [RFC-0202: TariScript Opcodes](RFC-0202_TariScriptOpcodes.md)

$$
\newcommand{\preimage}{\\phi} % pre image
\newcommand{\hash}[1]{\mathrm{H}\bigl({#1}\bigr)}
$$

## Description

> Atomic swaps are atomic transactions that allow users to exchange different crypto assets and or coins without the use of a central exchange. This makes it much more private and secure to do and swap because no third party is required for this to be secure. Atomic swaps work on the principle of using [Hashed Time Lock Contract](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts) or otherwise known as HTLCs. In short, it requires some pre-image(typically a hash, hence the name) to unlock the contract or the time part kicks in. 

Reviewer 3: ```suggestion
Atomic swaps are atomic transactions that allow users to trustlessly exchange different crypto assets and or coins without the use of a central exchange. This makes it much more private and secure to do and swap because no third party is required for this to be secure. Atomic swaps work on the principle of using [Hashed Time Lock Contract](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts)(HTLC). In short, it requires some hash pre-image to unlock the contract or the time-lock can be used to reclaim the funds. 
```

> In an Atomic swap, both users lock up the funds on their respective chains in an HTLC type contract. But the pre-image or spending knowledge of the two contracts are the same, but only one party knows the correct pre-image. When the first HTLC contract is spent, then this publicly displays the pre-image for the other party to use in spending the second HTLC. If the first HTLC is never spent the second time lock part kicks in allowing the user to respend the funds back to themselves.

Reviewer 3: ```suggestion
In an Atomic swap, both users lock up the funds to be exchanged on their respective chains in an HTLC-type contract. However, the pre-image, or spending secret, of the two contracts is the same, but only one party knows the correct pre-image. When the first HTLC contract is spent, then this publicly reveals the pre-image for the other party to use in spending the second HTLC. If the first HTLC is never spent the second transactions time-lock will allow the user to respend the funds back to themselves after a time.
```

### BTC - XTR AtomicSwap

#### Overview

> BTC uses a scripting language to enable smart contracts on transactions which provides [atomic swaps](https://tlu.tarilabs.com/protocols/atomic-swaps/AtomicSwaps.html). Traditionally [Mimblewimble] coins such as Tari does not have any scripting language, this makes Atomic swaps harder to implement, but has not been impossible, [mimblewimble atomic swaps](https://tlu.tarilabs.com/protocols/grin-protocol-overview/MainReport.html#atomic-swaps) but Tari does have scripting with [TariScript] which works a lot like BTC scripts, this makes adding in scripts a lot easier. Because of the same type of scripting, the scripts to both HTLCs will look very similar and we only need to ensure that we use the same Hashing function in both. 

Reviewer 3: ```suggestion
BTC uses a scripting language to enable smart contracts on transactions which enables [atomic swaps](https://tlu.tarilabs.com/protocols/atomic-swaps/AtomicSwaps.html) on the BTC chain. Traditionally [Mimblewimble] coins do not implement scripts, this makes Atomic swaps harder to implement but not impossible. Grin has implemented atomic swaps as a version of a 2-of-2 multi-signature transaction [mimblewimble atomic swaps](https://tlu.tarilabs.com/protocols/grin-protocol-overview/MainReport.html#atomic-swaps). Fortunately, Tari does have scripting with [TariScript] which works a lot like BTC scripts, this makes adding in scripts a lot easier. Because of the same type of scripting, the scripts to both HTLCs will look very similar and we only need to ensure that we use the same Hashing function in both. 
```

> To do an Atomic swap from BTC to XTR we need 4 wallets, 2 BTC wallets, and 2 XTR wallets. To make life easier, we can control most of this from the XTR wallets and using a GRPC or JsonRPC interface control the BTC wallets from the XTR wallets. 

Reviewer 3: The RPC stuff is implementation details that do not belong in the RFC.

> Let us explain the process in an example with Alice wanting to trade some of her BTC for Bob's XTR. Alice and Bob need to agree on an amount of XTR and BTC to swap. If this is agreed on the following steps will happen:

Reviewer 3: ```suggestion
As an example, Alice wants to trade some of her BTC for Bob's XTR. Alice and Bob need to agree on an amount of XTR and BTC to swap. If this is agreed on the following steps will happen:
```

> * Alice chooses a random string as the pre-image (\\( \preimage \\)), she hashes this with SHA256. She then sends this to Bob. Along with a public version of the [script key], \\( K_{Sa} \\) for use in the XTR transaction.

Reviewer 3: ```suggestion
* Alice chooses a random string as the pre-image (\\( \preimage \\)), she hashes this with SHA256. She then sends the hash to Bob. Along with a public version of the [script key], \\( K_{Sa} \\) for use in the XTR transaction.
```

* Bob sends his BTC address to Alice.

* Alice creates a BTC transaction with an HTLC contract requiring \\( \preimage \\) as the input, to pay out to Bob's address or it pays back to her after a certain time. 

> * Bob waits for this transaction to be mined. When it is mined, he checks the spending script and verifies that script checks (\\ \hash{\preimage} \\), sends it to his BTC address and that it has a sufficiently long timelock to give him time to claim the transaction.

Reviewer 3: ```suggestion
* Bob waits for this transaction to be mined. When it is mined, he checks the spending script and verifies that the script checks (\\ \hash{\preimage} \\), spends the funds to his BTC address and that it has a sufficiently long time-lock to give him time to claim the transaction.
```

* Alice checks the Tari blockchain and upon seeing that the transaction is mined, she claims the transaction. But by claiming it, she makes public what (\\( \preimage \\)) is.

* Bob sees that his XTR is spent, and looks at the input stack to get \\( \preimage \\). Bob can then use \\( \preimage \\) to claim the BTC transaction.

#### BTC HTLC script 

Here is the required BTC script that Bob publishes:

```
	OP_IF
	   OP_SHA256 <HASH256{pre_image}> OP_EQUALVERIFY
		<Alice BTC address> OP_CHECKSIG
	OP_ELSE
      <relative locktime>
      OP_CHECKSEQUENCEVERIFY
      OP_DROP
      <Alice BTC address> OP_CHECKSIG
   OP_ENDIF
```
_relative locktime_ is a time sequence that Alice chooses to lock up the funds in order to give Bob time to claim this. 

#### XTR - HTLC script 

Here is the required XTR script that Alice publishes:

```
   HashSha256 PushHash(HASH256{pre_image}) Equal
   IFTHEN
      PushPubkey(K_{Sa})
	ELSE
      CheckHeightVerify(height)
      PushPubkey(K_{Sb})
   ENDIF
```
(\\( K_{Sb} \\)) is the public key of the [script key] pair that Bob chooses to be able to claim this transaction if Alice backs out. 

> _height_ is a relative block height that Bob chooses to lock up the funds in order to give Alice time to claim the funds. 

Reviewer 3: Is it relative? I thought the heights are absolute?

> #### Communication

Reviewer 3: Implementation details, take this out

### BTC - XTR AtomicSwap

#### Overview

BTC uses a scripting language for smart contracts on transactions which enables [atomic swaps](https://tlu.tarilabs.com/protocols/atomic-swaps/AtomicSwaps.html) on the BTC chain. Traditionally [Mimblewimble] coins do not implement scripts, which makes Atomic swaps harder to implement but not impossible. Grin has implemented atomic swaps using a version of a 2-of-2 multi-signature transaction [mimblewimble atomic swaps](https://tlu.tarilabs.com/protocols/grin-protocol-overview/MainReport.html#atomic-swaps). Fortunately, Tari does have scripting with [TariScript] which works a lot like BTC scripts, which make the implementation simpler. Because of the scripting similarities, the scripts to both HTLCs will look very similar and we only need to ensure that we use the same hash function in both. 

To do an Atomic swap from BTC to XTR we need 4 wallets, 2 BTC wallets, and 2 XTR wallets, 1 wallet per person, per coin.

As an example, Alice wants to trade some of her XTR for Bob's BTC. Alice and Bob need to agree on an amount of XTR and BTC to swap. Once an agreement is reached, the swap is executed in the following steps:

> * Alice chooses a random string, \\( \preimage \\), as the pre-image and hashes it with SHA256. She then sends the hash 

Reviewer 3: ```suggestion
* Alice chooses a set of random bytes, \\( \preimage \\), as the pre-image and hashes it with SHA256. She then sends the hash 
```

Just wanted to clarify that a pre-image is not necessarily a string, it is random data to act as entropy.

      OP_CHECKSEQUENCEVERIFY
      OP_DROP
      <Bob BTC address> OP_CHECKSIG
   OP_ENDIF
```
_relative locktime_ is a time sequence that Alice chooses to lock up the funds in order to give Bob time to claim this. 

#### XTR - HTLC script 

Here is the required XTR script that Alice publishes:

```
   HashSha256 PushHash(HASH256{pre_image}) Equal
   IFTHEN
      PushPubkey(K_{Sb})
	ELSE
      CheckHeightVerify(height)
      PushPubkey(K_{Sa})
   ENDIF
```
(\\( K_{Sb} \\)) is the public key of the [script key] pair that Bob chooses to be able to claim this transaction if Alice backs out. 
_height_ is a absolute block height that Bob chooses to lock up the funds in order to give Alice time to claim the funds. 

## Notation

Where possible, the "usual" notation is used to denote terms commonly found in cryptocurrency literature. Lower case 
characters are used as private keys, while uppercase characters are used as public keys. New terms introduced here are assigned greek lowercase letters in most cases. Some terms used here are noted down in [TariScript]. 

| Name        | Symbol              | Definition |
|:------------|---------------------| -----------|

> | Pre-image   | \\( \preimage \\) | The random string used for the pre-image of the hash |

Reviewer 3: ```suggestion
| Pre-image   | \\( \preimage \\) | The random byte data used for the pre-image of the hash |
```
